components/script/dom/vr.rs

```diff
@@ -0,0 +1,173 @@
+                    // Sync displays
+                    for display in displays {
+                        self.sync_display(&display);
+                    }
+                },
+                Err(e) => {
+                    promise.reject_native(promise.global().get_cx(), &e);
+                    return promise;
+                }
+            }
+        }
+
+        // convert from JS to Root
+        let displays: Vec<Root<VRDisplay>> = self.displays.borrow().iter()
+                                                          .map(|d| Root::from_ref(&**d))
+                                                          .collect();
+        promise.resolve_native(promise.global().get_cx(), &displays);
+
+        promise
+    }
+
+    // https://w3c.github.io/webvr/#interface-navigator
+    fn VrEnabled(&self) -> bool {
+        // TODO: check iframe
+        true
+    }
+}
+
+
+impl VR {
+    fn webvr_thread(&self) -> Option<IpcSender<WebVRMsg>> {
+        self.global().as_window().webvr_thread()
+    }
+
+    fn find_display(&self, display_id: u64) -> Option<Root<VRDisplay>> {
+        self.displays.borrow()
+                     .iter()
+                     .find(|d| d.get_display_id() == display_id)
+                     .map(|d| Root::from_ref(&**d))
+    }
+
+    fn register(&self) {
+        if let Some(wevbr_sender) = self.webvr_thread() {
+             let msg = WebVRMsg::RegisterContext(self.global().pipeline_id());
+             wevbr_sender.send(msg).unwrap();
+        }
+    }
+
+    fn unregister(&self) {
+        if let Some(wevbr_sender) = self.webvr_thread() {
+             let msg = WebVRMsg::UnregisterContext(self.global().pipeline_id());
+             wevbr_sender.send(msg).unwrap();
+        }
+    }
+
+    fn sync_display(&self, display: &webvr::VRDisplayData) {
+        if let Some(existing) = self.find_display(display.display_id) {
+            existing.update_display(&display);
+        } else {
+            let root = VRDisplay::new(&self.global(), &display);
+            self.displays.borrow_mut().push(JS::from_ref(&*root));
+        }
+    }
+
+    pub fn handle_webvr_event(&self, event: WebVREventMsg) {
+        let event = match event {
+            WebVREventMsg::DisplayEvent(display_event) => display_event
+        };
+
+        match &event {
+            &webvr::VRDisplayEvent::Connect(ref display) => {
+                self.sync_display(&display);
+                if let Some(display) = self.find_display(display.display_id) {
+                    display.handle_webvr_event(&event);
+                    self.notify_event(&display, &event);
+                }
+            },
+            &webvr::VRDisplayEvent::Disconnect(id) => {
+                if let Some(display) = self.find_display(id) {
+                    display.handle_webvr_event(&event);
+                    self.notify_event(&display, &event);
+                }
+            },
+            &webvr::VRDisplayEvent::Activate(ref display, _) |
```

Comment:
Oh I see, this is because you borrow event afterwards, right?

Then you can do something like:

```rust
ref event @ webvr::VRDisplayEvent::Deactivate(ref display) |
// ...
```

Though I'm not sure that'd be a strict improvement.